[lldb/crashlog] Make interactive mode the new default

medismailben opened the pull request:
This patch makes interactive mode as the default when using the crashlog command. It replaces the existing `-i|--interactive` flag with a new `-m|--mode` option, that can either be `interactive` or `batch`.

By default, when the option is not explicitely set by the user, the interactive mode is selected, however, lldb will fallback to batch mode if the command interpreter is not interactive or if stdout is not a tty.

This also adds some railguards to prevent users from using interactive only options with the batch mode and updates the tests accordingly.

rdar://97801509

Differential Revision: https://reviews.llvm.org/D141658

medismailben commented:
FWIW, this was already approved in https://reviews.llvm.org/D141658 but I've updated it slightly to match to current version of the crashlog script.

bulbazord commented on lldb/test/Shell/ScriptInterpreter/Python/Crashlog/skipped_status_interactive_crashlog.test:
```diff
@@ -3,7 +3,7 @@
 # RUN: yaml2obj %S/Inputs/interactive_crashlog/multithread-test.yaml > %t.dir/multithread-test
 # RUN: %lldb -b -o 'command script import lldb.macosx.crashlog' \
-# RUN: -o 'crashlog -a -i -s -t %t.dir/multithread-test %S/Inputs/interactive_crashlog/multithread-test.ips' \
+# RUN: -o 'crashlog -a -s --mode interactive -t %t.dir/multithread-test %S/Inputs/interactive_crashlog/multithread-test.ips' \
```
If interactive mode is the new default, should we drop the `--mode interactive` and have the test rely on the default behavior? :)

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -1794,8 +1803,36 @@ def SymbolicateCrashLogs(debugger, command_args, result, is_command):
+        and (options.target_path or options.skip_status)
+    ):
+        print(
+            "Target path (-t) and skipping process status (-s) options can only used in intercative mode (-m=interactive)."
```
typo: `intercative` -> `interactive`

medismailben commented:
Closed this by mistake.

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -1794,8 +1803,36 @@ def SymbolicateCrashLogs(debugger, command_args, result, is_command):
+        options.mode = CrashLogLoadingMode.batch
+
+    if (
+        options.mode
```
Isn't `options.mode` always set to interactive? It defaults to that up in the argparse code, so is this check necessary?

medismailben replied:
Oversight.

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -1821,31 +1858,24 @@ def SymbolicateCrashLogs(debugger, command_args, result, is_command):
 
     def should_run_in_interactive_mode(options, ci):
-        if options.interactive:
+        if options.mode:
```
When is this ever false?

medismailben commented on lldb/examples/python/crashlog.py:
```diff
@@ -1794,8 +1803,36 @@ def SymbolicateCrashLogs(debugger, command_args, result, is_command):
+    if options.interactive:
+        options.mode = CrashLogLoadingMode.interactive
+    elif options.batch:
+        options.mode = CrashLogLoadingMode.batch
```
You can't do `-i -b` (or vice versa) because they're in a mutually exclusive group.

To answer the first part of the question, we set `options.mode` here because it's the main "source of truth", so if you did `--mode interactive -b` we would technically override the mode to batch.

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -1794,8 +1802,35 @@ def SymbolicateCrashLogs(debugger, command_args, result, is_command):
+        if options.batch:
+            options.mode = CrashLogLoadingMode.batch
+        else:
+            options.mode = CrashLogLoadingMode.interactive
```
Can you detect both `options.batch` and `options.interactive` being specified at the same time? I assume whichever is listed first will get selected... but I would assume we'd want to add something that prevents you from choosing both.

medismailben replied:
It's illegal to have both `batch` & `interactive` and `argparse` will prevent you from doing anything.

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -1821,31 +1856,24 @@ def SymbolicateCrashLogs(debugger, command_args, result, is_command):
 
     def should_run_in_interactive_mode(options, ci):
-        if options.interactive:
+        if options.mode == CrashLogLoadingMode.interactive:
```
I guess I'm somewhat confused by the purpose of this function... The only way `options.mode` is ever not interactive is if it's explicitly set by the user to be something else right? Why might we want to ignore that?

bulbazord reviewed (commented):
LGTM in general, a few more comments but otherwise this is looking pretty good

medismailben committed:
[lldb/crashlog] Make interactive mode the new default

This patch makes interactive mode as the default when using the crashlog
command. It replaces the existing `-i|--interactive` flag with a new
`-m|--mode` option, that can either be `interactive` or `batch`.

By default, when the option is not explicitely set by the user, the
interactive mode is selected, however, lldb will fallback to batch
mode if the command interpreter is not interactive or if stdout is not
a tty.

This also adds some railguards to prevent users from using interactive
only options with the batch mode and updates the tests accordingly.

rdar://97801509

Differential Revision: https://reviews.llvm.org/D141658

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
